sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryParallelReadTableSource.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.sdk.io.gcp.bigquery;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.api.services.bigquery.model.Table;
+import com.google.api.services.bigquery.model.TableReference;
+import com.google.cloud.bigquery.v3.ParallelRead.CreateSessionRequest;
+import com.google.cloud.bigquery.v3.ParallelRead.ReadLocation;
+import com.google.cloud.bigquery.v3.ParallelRead.Session;
+import com.google.cloud.bigquery.v3.ReadOptions;
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.io.BoundedSource;
+import org.apache.beam.sdk.io.gcp.bigquery.BigQueryHelpers.TableRefToJson;
+import org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO.ReadSessionOptions;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.options.ValueProvider;
+import org.apache.beam.sdk.options.ValueProvider.NestedValueProvider;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An abstract {@link BoundedSource} to read a table from BigQuery.
+ *
+ * <p>This source is to read Bigquery data directly using ParallelRead API.
+ * <ul>
+ * <li>{@link BigQueryParallelReadTableSource} is for reading BigQuery tables</li>
+ * </ul>
+ * ...
+ */
+class BigQueryParallelReadTableSource<T> extends BoundedSource<T> {
+  private static final Logger LOG = LoggerFactory.getLogger(BigQueryParallelReadTableSource.class);
+
+  private BigQueryServices bqServices;
```

Comment:
All members should be final (except for transient ones) - sources are immutable

Reply:
Done. Per offline discussion, we create a read session as a transient object in split() -- we cache the split results today, but in the future we might want to throw away the old read session and create a new one in order to accommodate a smaller chunk size (e.g. when we support sub-file sharding). We think this is OK since read sessions are cheap and are self-deleting.